File: packages/capacitor-plugin/ios/Sources/LocalLLMPlugin/LocalLLMPlugin.swift
import Foundation
import Capacitor

@objc(LocalLLMPlugin)
public class LocalLLMPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "LocalLLMPlugin"
    public let jsName = "LocalLLM"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "systemAvailability", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "prompt", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "endSession", returnType: CAPPluginReturnPromise)
        
    ]
    
    private let implementation = LocalLLM()

    @objc func systemAvailability(_ call: CAPPluginCall) {
        call.resolve([
            "status": LocalLLM.availability().rawValue
        ])
    }
    
    @objc func prompt(_ call: CAPPluginCall) {
        Task {
            let options = getLLMPromptOptionsFromCall(call)
            
            do {
                let responseText = try await self.implementation.prompt(options: options)
                
                call.resolve([
                    "text": responseText
                ])
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }
    
    @objc func endSession(_ call: CAPPluginCall) {
        guard let sessionId = call.getString("sessionId") else {
            call.reject("sessionId is required")
            return
        }
        
        implementation.endSession(sessionId)
        
        call.resolve()
    }
    
    private func getLLMPromptOptionsFromCall(_ call: CAPPluginCall) -> LLMPromptOptions {
        return LLMPromptOptions(
            sessionId: call.getString("sessionId"),
            instructions: call.getString("instructions"),
            options: getLLMOptionsFromCall(call),
            prompt: call.getString("prompt", "")
        )
    }
    
    private func getLLMOptionsFromCall(_ call: CAPPluginCall) -> LLMOptions? {
        guard let optionsObject = call.getObject("options") else {
            return nil
        }
        
        return LLMOptions(
            temperature: optionsObject["temperature"] as? Double,
            maximiumOutputTokens: optionsObject["maximiumOutputTokens"] as? Int,
        )
    }
}

File: packages/capacitor-plugin/android/src/main/java/io/ionic/localllm/plugin/LocalLLMPlugin.kt
package io.ionic.localllm.plugin

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.runBlocking

@CapacitorPlugin(name = "LocalLLM")
class LocalLLMPlugin : Plugin() {
  private var implementation: LocalLLM? = null

    override fun load() {
        super.load()
        implementation = LocalLLM(context)
    }

  @PluginMethod fun systemAvailability(call: PluginCall) {
      try {
          val impl = this.implementation ?: throw Exception("LocalLLM not initialized")
          call.resolve(JSObject().put("status", impl.availability().value))
      } catch (ex: Exception) {
          call.reject(ex.message)
      }
  }

  @PluginMethod
  fun prompt(call: PluginCall) {
      runBlocking {
          try {
              val options = getLLMPromptOptionsFromCall(call)

              val impl = this@LocalLLMPlugin.implementation ?: throw Exception("LocalLLM not initialized")
              val response = impl.prompt(options)
              call.resolve(JSObject().put("text", response))
          } catch (ex: Exception) {
              call.reject(ex.message)
          }
      }
  }

    private fun getLLMPromptOptionsFromCall(call: PluginCall): LLMPromptOptions {
        return LLMPromptOptions(
            sessionId = call.getString("sessionId"),
            instructions = call.getString("instructions"),
            options = getLLMOptionsFromCall(call),
            prompt = call.getString("prompt")!!
        )
    }

    private fun getLLMOptionsFromCall(call: PluginCall): LLMOptions? {
        val optionsObject = call.getObject("options")

        if (optionsObject!= null) {
            return LLMOptions(
                temperature = optionsObject.optDouble("temperature")?.toFloat(),
                maxOutputTokens =  optionsObject.optInt("maximiumOutputTokens")
            )
        }

        return null
    }
}
